salt/grains/core.py

```diff
@@ -984,6 +984,60 @@ def _virtual(osdata):
     return grains
 
 
+def _virtual_hv(osdata):
+    '''
+    Returns detailed hypervisor information from sysfs
+    Currently this seems to be used only by Xen
+    '''
+    grains = {}
+
+    # Bail early if we're not running on Xen
+    if 'xen' not in osdata['virtual']:
+        return grains
+
+    # Try to get the exact hypervisor version from sysfs
+    try:
+        version = {}
+        for fn in ('major', 'minor', 'extra'):
+            with salt.utils.files.fopen('/sys/hypervisor/version/{}'.format(fn), 'r') as fhr:
+                version[fn] = fhr.read().strip()
```

Comment:
This needs to be decoded to unicode using:

```python
version[fn] = salt.utils.stringutils.to_unicode(fhr.read().strip())
```

This may also require that `salt.utils.stringutils` is added to the imports.